### ISSUE TITLE
[BUG] - Longhorn instance-manager only uses max 2 CPU

### ISSUE THREAD

**Issue by pwurbs** [longhorn/longhorn#9722]
## Describe the bug

There is an app with rather high IO (500-1000 IOPS). If there is a Longhorn background task (like replica rebuild or snapshot deletion), then the instance-manager pod running at the relevant worker node is under heavy load and consumes 1-2 CPU. 
There are 2 issues:
- The background task is quite slow. E.g. snapshot deletion (merge 250GB into 1.8TB) takes 15hrs. For the same snapshot deletion on another node without IO intense applications only 1hr is needed
- IO intense application suffers due to overloaded instance-manager, IO utilization for the logical Longhorn /dev/sdm device shows nearly 100%. 

The instance-manager pod seems to be the bottleneck. The actual node disk doesn't seem to be the bottleneck (about 20% utilization).
The worker node provides 16 CPU overall. Max 2 CPUs are used by the instance-manager, max 1 CPU by other pods, 13 CPU are mostly free.
**Now the bug:**
If the instance-manager seems to be so busy (leading to over utilization of the logical sdm device and quite slow background tasks), why doesn't the instance-manager pod use more than 2 CPU? The CPU request value (danger zone) is kept as default (1.92 CPU), no K8S limit configured. 
Is there any limitation within the instance-manager software not being able to scale to use more CPU?

Screenshot shows the CPU graph of the instance-manager pod.
Increased CPU load starts at about 05:00 when the snapshot merge started. No more than 2 CPU is used
![image](https://github.com/user-attachments/assets/263b11d1-caa2-49f4-bf36-3473adf044a5)

This screenshot shows the disk utilization over time:
logical "sdm" device nearly 100%
"physical" node device "vdc" about 20%
![image](https://github.com/user-attachments/assets/edd6396a-70d6-42cd-8ee6-1d283d611edc)

## To Reproduce

- Deploy an IO-intense application
- Start a Longhorn background task
- Watch the instance-manager pod metrics
- Watch the slow progress of background tasks

## Expected behavior

instance-manager should use more than 2 CPU when required and if available on the worker node

## Environment

 - Longhorn version: 1.5.5
 - Impacted volume (PV): 
 - Installation method (e.g. Rancher Catalog App/Helm/Kubectl): Helm
 - Kubernetes distro (e.g. RKE/K3s/EKS/OpenShift) and version: RKE
   - Number of control plane nodes in the cluster: 3
   - Number of worker nodes in the cluster: 6
 - Node config
   - OS type and version: AlmaLinux 8.10 (Cerulean Leopard)
   - Kernel version: 4.18.0-553.22.1.el8_10.x86_64
   - CPU per node: 16
   - Memory per node: 64GB
   - Disk type (e.g. SSD/NVMe/HDD): SSD
   - Network bandwidth between the nodes (Gbps): 10Gbps
 - Underlying Infrastructure (e.g. on AWS/GCE, EKS/GKE, VMWare/KVM, Baremetal): Cloud
 - Number of Longhorn volumes in the cluster: 49

**Comment by DamiaSan** [comment]
Have you tried to change this setting? https://longhorn.io/docs/1.7.2/references/settings/#guaranteed-instance-manager-cpu
Here you can also find some references: https://github.com/longhorn/longhorn/blob/v1.7.2/scalability/reference-setup-performance-scalability-and-sizing-guidelines/public-cloud/medium-node-spec.md

**Comment by pwurbs** [comment]
@DamiaSan 
No, I didn't change the "Guaranteed Instance Manager CPU" setting, because it only defines the K8S request value. So this should not have a direct influence how many CPUs are actually used. A K8S limit is not defined. There is also no K8S CPU throttling.
So, if a pod has only a request specified and no limit, it could theoretically use more CPU. 
Or is there any internal magic function within the instance-manager software which uses the request value to define how many CPU the software shall use or influences the internal thread logic?
The 2nd link doesn't give configuration advice.

**Comment by DamiaSan** [comment]
cc @PhanLe1010

**Comment by pwurbs** [comment]
Additional observation:
- Very slow replica rebuild of a 2TB volume, already lasting for 20hrs, only 35% achieved
- Node has 16 CPU, load is at 5 
- No IOWait, no higher disk device utilization, no network saturation
- Very little network and IO activity
- instance-manager pod on the source node constantly consumes 2CPU
- instance-manager on target node is mostly idle
Questions:
- What does the instance-manager do over such long time, using so much CPU, if there is fairly no progress for rebuild
- If the instance-manager does anything useful, why doesn't it consume more than 2 CPU?

**Comment by innobead** [comment]
It's helpful to continue investigating this issue with @pwurbs to clarify the context. removed the stale label.

cc @shuo-wu @derekbit

**Comment by shuo-wu** [comment]
Thank you for the feedback. Typically users will complain that longhorn manager or instance manager consumes too many CPUs. The case you are reporting is relatively rare. I am not sure if there can be any default limitations/configurations in your k8s cluster or linux (like cgroup). For now, I have two questions about the issue:
1. Is the logical sdm device a Longhorn volume? Or the Longhorn data path on this worker node is based on this logical device.
2. Have you tried to directly put a heavy IO load to a Longhorn volume without any Longhorn background task (like rebuilding, snapshot deletion, etc)?

**Comment by pwurbs** [comment]
Here are the answers to your questions:
1. Now, the LH device is sda and the volume name is pvc-978e2c01-cf30-4372-b1a4-5d508c58877a
sda is the device name where the volume is attached
lsblk returns:
`sda      8:0    0  2.6T  0 disk /var/lib/kubelet/pods/87349c77-10e4-42f7-93c9-dbf5b83d5e93/volumes/kubernetes.io~csi/pvc-978e2c01-cf30-4372-b1a4-5d508c58877a/mount`
mount returns:
`/dev/longhorn/pvc-978e2c01-cf30-4372-b1a4-5d508c58877a on /var/lib/kubelet/plugins/kubernetes.io/csi/driver.longhorn.io/d3238fc2966120c4b18639c63b36b56266f5d7693758e18fb0cb0a166d89a389/globalmount type ext4 (rw,relatime,seclabel)`
3. We didn't test this consciously. We see peaks per workload (container) up to 3000 read and 500 write IOPS. Throughput is not more than 30MBps. Per node, there are write IOPS peaks up to 3000 and throughput up to 300MBps (Read)

We only face this issue by selected volumes. Most of the volumes replicate with sufficient performance.

Now we have the same issue with another volume too. It's even more stuck. Here is some information:
- Volume name pvc-a4191518-32c0-4de2-abb2-57b4e47e26d2
- Rebuild source: worker01, instance-manager-fb4f06826cf4b7304d2ccb773053e58f
- Rebuild target: worker03, instance-manager-575c38b27a32a39d7e29622d0051aed5
- Status worker01: load=4, network only 4MBps, physical disk utilization peak 15%
- Status worker03: load=0.4, network only 4MBps, physical disk utilization peak 11%
- Status instance-manager-fb4f06826cf4b7304d2ccb773053e58f (source):  2 CPU sustained over at least 7 days. Network only 1-15MBps transmit
- Status instance-manager-575c38b27a32a39d7e29622d0051aed5 (target):  0.1-0.2 CPU sustained over 1 day, some peaks up to 1 CPU. Network only 1-15MBps receive

--> Sending instance-manager consumes 2 CPU sustained, but there is fairly no rebuild progress (and very low network transmission), network and node is not overloaded at all

In the meanwhile, we use LH 1.6.3.

**Comment by pwurbs** [comment]
We created a new volume and copied all the data into it. Now the status is fine: Fast replica rebuild, fast backup etc.
So, there must be any kind of "inconsistency" of a volume, even if it's marked as healthy. In our case we store many many files in it and the files are created and deleted with a rather high frequency. Is there a possibility that a volume gets fragmented under such conditions or anything like that?

**Comment by derekbit** [comment]
> We created a new volume and copied all the data into it. Now the status is fine: Fast replica rebuild, fast backup etc. So, there must be any kind of "inconsistency" of a volume, even if it's marked as healthy. In our case we store many many files in it and the files are created and deleted with a rather high frequency. Is there a possibility that a volume gets fragmented under such conditions or anything like that?

Your underlying disks are SSDs, so I think the ceiling of IO performance might be nothing to do with fragmentation.

> In our case we store many files in it and the files are created and deleted with a rather high frequency

Can you elaborate more on the IO behaviors? What sizes of the files? How many files? Deletion frequency?
If you can provide more information or even a script to simulate your IO behaviors, we can try to reproduce and investigate the bottleneck.

**Comment by pwurbs** [comment]
Also SSD has a limit and imposes some latency...
Some numbers for the affected volume:
- Size: 2TB
- 8 Mio files
- 7 Mio files are changed/replaced more often
I will try to find out the deletion frequency

**Comment by pwurbs** [comment]
It would be sufficient for me, if the Longhorn developers could at least give some information if there is any (implicit) CPU limit in code or logic.
